bundles/org.openhab.binding.juicenet/src/main/java/org/openhab/binding/juicenet/internal/handler/JuiceNetDeviceHandler.java

```diff
@@ -0,0 +1,349 @@
+
+    private final TimeZoneProvider timeZoneProvider;
+
+    // properties
+    private String name = "";
+
+    private String token = "";
+    private long targetTimeTou = 0;
+    private long lastInfoTimestamp = 0;
+
+    JuiceNetApiDeviceStatus deviceStatus = new JuiceNetApiDeviceStatus();
+    JuiceNetApiInfo deviceInfo = new JuiceNetApiInfo();
+    JuiceNetApiTouSchedule deviceTouSchedule = new JuiceNetApiTouSchedule();
+    JuiceNetApiCar deviceCar = new JuiceNetApiCar();
+
+    public JuiceNetDeviceHandler(Thing thing, TimeZoneProvider timeZoneProvider) {
+        super(thing);
+
+        this.timeZoneProvider = timeZoneProvider;
+    }
+
+    public void setNameAndToken(String name, String token) {
+        logger.trace("setNameAndToken");
+        this.token = token;
+
+        if (!name.equals(this.name)) {
+            updateProperty(PROPERTY_NAME, name);
+            this.name = name;
+        }
+
+        if (getThing().getStatus() != ThingStatus.ONLINE) {
+            goOnline();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        logger.trace("JuiceNetDeviceHandler:initialize");
+        Configuration configuration = getThing().getConfiguration();
+
+        String stringId = configuration.get(PARAMETER_UNIT_ID).toString();
+        if (stringId.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                    "@text/offline.configuration-error.id-missing");
+            return;
+        }
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // This device will go ONLINE on the first successful API call in queryDeviceStatusAndInfo
+    }
+
+    public void handleApiException(Exception e) {
+        if (e instanceof JuiceNetApiException) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.toString());
+        } else if (e instanceof InterruptedException) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.toString());
+            Thread.currentThread().interrupt();
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.NONE, e.toString());
+        }
+    }
+
+    public void goOnline() {
+        logger.trace("goOnline");
+        if (this.getThing().getStatus() == ThingStatus.ONLINE) {
+            return;
+        }
+
+        if (token.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                    "@text/offline.configuration-error.non-existent-device");
+            return;
+        }
+
+        try {
+            tryQueryDeviceStatusAndInfo();
+        } catch (JuiceNetApiException | InterruptedException e) {
+            handleApiException(e);
+            return;
+        }
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    protected JuiceNetApi getApi() {
+        Bridge bridge = Objects.requireNonNull(getBridge());
```

Comment:
This will throw if bridge is not configured. This could be avoided by validation in `initialize`:

```java
        Bridge bridge = getBridge();
        if (bridge == null) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
                    "@text/offline.configuration-error.bridge-missing");
            return;
        }
```